# Machine Learning Engineer Nanodegree
## Supervised Learning
## Project: Building a Student Intervention System

Welcome to the second project of the Machine Learning Engineer Nanodegree! In this notebook, some template code has already been provided for you, and it will be your job to implement the additional functionality necessary to successfully complete this project. Sections that begin with **'Implementation'** in the header indicate that the following block of code will require additional functionality which you must provide. Instructions will be provided for each section and the specifics of the implementation are marked in the code block with a `'TODO'` statement. Please be sure to read the instructions carefully!

In addition to implementing code, there will be questions that you must answer which relate to the project and your implementation. Each section where you will answer a question is preceded by a **'Question X'** header. Carefully read each question and provide thorough answers in the following text boxes that begin with **'Answer:'**. Your project submission will be evaluated based on your answers to each of the questions and the implementation you provide.  

>**Note:** Code and Markdown cells can be executed using the **Shift + Enter** keyboard shortcut. In addition, Markdown cells can be edited by typically double-clicking the cell to enter edit mode.

### Question 1 - Classification vs. Regression
*Your goal for this project is to identify students who might need early intervention before they fail to graduate. Which type of supervised learning problem is this, classification or regression? Why?*

**Answer: **
Since the possible outputs lies on a discrete universe (binary in this case - {intervention, not intervention}), the type of supervised learning problem is classification, seeing that it must classify (choose between a finite and discrete number of classes), as opposite to a regression problem (where the output lies on a continous universe).

## Exploring the Data
Run the code cell below to load necessary Python libraries and load the student data. Note that the last column from this dataset, `'passed'`, will be our target label (whether the student graduated or didn't graduate). All other columns are features about each student.

In [2]:
# Import libraries
import numpy as np
import pandas as pd
from time import time
from sklearn.metrics import f1_score

# Read student data
student_data = pd.read_csv("student-data.csv")
print "Student data read successfully!"

Student data read successfully!


### Implementation: Data Exploration
Let's begin by investigating the dataset to determine how many students we have information on, and learn about the graduation rate among these students. In the code cell below, you will need to compute the following:
- The total number of students, `n_students`.
- The total number of features for each student, `n_features`.
- The number of those students who passed, `n_passed`.
- The number of those students who failed, `n_failed`.
- The graduation rate of the class, `grad_rate`, in percent (%).


In [23]:
# TODO: Calculate number of students
n_students = len(student_data.axes[0])

# TODO: Calculate number of features - OBS: dropping passed as it is the output and not a feature
n_features = len((student_data.drop(columns=['passed']).columns))

# TODO: Calculate passing students
n_passed = len(student_data[student_data['passed'] == 'yes'])

# TODO: Calculate failing students
n_failed = len(student_data[student_data['passed'] == 'no'])

# TODO: Calculate graduation rate
grad_rate = (float(n_passed)/n_students)*100

# Print the results
print "Total number of students: {}".format(n_students)
print "Number of features: {}".format(n_features)
print "Number of students who passed: {}".format(n_passed)
print "Number of students who failed: {}".format(n_failed)
print "Graduation rate of the class: {:.2f}%".format(grad_rate)

Total number of students: 395
Number of features: 30
Number of students who passed: 265
Number of students who failed: 130
Graduation rate of the class: 67.09%


## Preparing the Data
In this section, we will prepare the data for modeling, training and testing.

### Identify feature and target columns
It is often the case that the data you obtain contains non-numeric features. This can be a problem, as most machine learning algorithms expect numeric data to perform computations with.

Run the code cell below to separate the student data into feature and target columns to see if any features are non-numeric.

In [24]:
# Extract feature columns
feature_cols = list(student_data.columns[:-1])

# Extract target column 'passed'
target_col = student_data.columns[-1] 

# Show the list of columns
print "Feature columns:\n{}".format(feature_cols)
print "\nTarget column: {}".format(target_col)

# Separate the data into feature data and target data (X_all and y_all, respectively)
X_all = student_data[feature_cols]
y_all = student_data[target_col]

# Show the feature information by printing the first five rows
print "\nFeature values:"
print X_all.head()

Feature columns:
['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']

Target column: passed

Feature values:
  school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

    ...    higher internet  romantic  famrel  freetime goout Dalc Walc health  \
0   ...       yes       no        no       4         3     4    1    1      3   
1   ...       

### Preprocess Feature Columns

As you can see, there are several non-numeric columns that need to be converted! Many of them are simply `yes`/`no`, e.g. `internet`. These can be reasonably converted into `1`/`0` (binary) values.

Other columns, like `Mjob` and `Fjob`, have more than two values, and are known as _categorical variables_. The recommended way to handle such a column is to create as many columns as possible values (e.g. `Fjob_teacher`, `Fjob_other`, `Fjob_services`, etc.), and assign a `1` to one of them and `0` to all others.

These generated columns are sometimes called _dummy variables_, and we will use the [`pandas.get_dummies()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html?highlight=get_dummies#pandas.get_dummies) function to perform this transformation. Run the code cell below to perform the preprocessing routine discussed in this section.

In [25]:
def preprocess_features(X):
    ''' Preprocesses the student data and converts non-numeric binary variables into
        binary (0/1) variables. Converts categorical variables into dummy variables. '''
    
    # Initialize new output DataFrame
    output = pd.DataFrame(index = X.index)

    # Investigate each feature column for the data
    for col, col_data in X.iteritems():
        
        # If data type is non-numeric, replace all yes/no values with 1/0
        if col_data.dtype == object:
            col_data = col_data.replace(['yes', 'no'], [1, 0])

        # If data type is categorical, convert to dummy variables
        if col_data.dtype == object:
            # Example: 'school' => 'school_GP' and 'school_MS'
            col_data = pd.get_dummies(col_data, prefix = col)  
        
        # Collect the revised columns
        output = output.join(col_data)
    
    return output

X_all = preprocess_features(X_all)
print "Processed feature columns ({} total features):\n{}".format(len(X_all.columns), list(X_all.columns))

Processed feature columns (48 total features):
['school_GP', 'school_MS', 'sex_F', 'sex_M', 'age', 'address_R', 'address_U', 'famsize_GT3', 'famsize_LE3', 'Pstatus_A', 'Pstatus_T', 'Medu', 'Fedu', 'Mjob_at_home', 'Mjob_health', 'Mjob_other', 'Mjob_services', 'Mjob_teacher', 'Fjob_at_home', 'Fjob_health', 'Fjob_other', 'Fjob_services', 'Fjob_teacher', 'reason_course', 'reason_home', 'reason_other', 'reason_reputation', 'guardian_father', 'guardian_mother', 'guardian_other', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences']


### Implementation: Training and Testing Data Split
So far, we have converted all _categorical_ features into numeric values. For the next step, we split the data (both features and corresponding labels) into training and test sets. In the following code cell below, you will need to implement the following:
- Randomly shuffle and split the data (`X_all`, `y_all`) into training and testing subsets.
  - Use 300 training points (approximately 75%) and 95 testing points (approximately 25%).
  - Set a `random_state` for the function(s) you use, if provided.
  - Store the results in `X_train`, `X_test`, `y_train`, and `y_test`.

In [40]:
# TODO: Import any additional functionality you may need here
import random
# TODO: Set the number of training points
num_train = 300

# Set the number of testing points
num_test = X_all.shape[0] - num_train

# TODO: Shuffle and split the dataset into the number of training and testing points above
seed = 0
random.seed(seed)
range_idx = range(0, X_all.shape[0])
shuffled_idx = random.sample(range_idx, X_all.shape[0])
X_train = X_all.iloc[shuffled_idx[0:num_train]]
X_test = X_all.iloc[shuffled_idx[num_train:]]
y_train = y_all.iloc[shuffled_idx[0:num_train]]
y_test = y_all.iloc[shuffled_idx[num_train:]]

# Show the results of the split
print "Training set has {} samples.".format(X_train.shape[0])
print "Testing set has {} samples.".format(X_test.shape[0])

Training set has 300 samples.
Testing set has 95 samples.


## Training and Evaluating Models
In this section, you will choose 3 supervised learning models that are appropriate for this problem and available in `scikit-learn`. You will first discuss the reasoning behind choosing these three models by considering what you know about the data and each model's strengths and weaknesses. You will then fit the model to varying sizes of training data (100 data points, 200 data points, and 300 data points) and measure the F<sub>1</sub> score. You will need to produce three tables (one for each model) that shows the training set size, training time, prediction time, F<sub>1</sub> score on the training set, and F<sub>1</sub> score on the testing set.

**The following supervised learning models are currently available in** [`scikit-learn`](http://scikit-learn.org/stable/supervised_learning.html) **that you may choose from:**
- Gaussian Naive Bayes (GaussianNB)
- Decision Trees
- Ensemble Methods (Bagging, AdaBoost, Random Forest, Gradient Boosting)
- K-Nearest Neighbors (KNeighbors)
- Stochastic Gradient Descent (SGDC)
- Support Vector Machines (SVM)
- Logistic Regression

### Question 2 - Model Application
*List three supervised learning models that are appropriate for this problem. For each model chosen*
- Describe one real-world application in industry where the model can be applied. *(You may need to do a small bit of research for this — give references!)* 
- What are the strengths of the model; when does it perform well? 
- What are the weaknesses of the model; when does it perform poorly?
- What makes this model a good candidate for the problem, given what you know about the data?

**Answer: **

*Reference from scikit-learn documentation:*
![Machine Learning Map](http://scikit-learn.org/stable/_static/ml_map.png)

First, lets analyse our dataset and its characteristics:
- Number of classes: just two (binary);
- Number of features: relatively high compared to the number of samples (48);
- Number of samples: 395, low size dataset

It seems our chosen models might overcome possible overfitting (must be robust), deal well with the high dimensionality (too much features) and perform fine under a low samples dataset. We might not worry too much with training time, since our dataset has so few samples.
From this characteristics, it seems reasonable to choose the following models from scikit-learn: Support Vector Machines, Ensemble method (Adaboost) and Logistic Regression.

 - *List three supervised learning models that are appropriate for this problem.*
  - *Support Vector Machines, Ensemble method (Adaboost) and Logistic Regression.*

**Support Vector Machines (SVM)**
- Describe one real-world application in industry where the model can be applied.
  - *Predicting diabetes and pre-diabetes in pacients: https://bmcmedinformdecismak.biomedcentral.com/articles/10.1186/1472-6947-10-16*


- What are the strengths of the model; when does it perform well? 
  - *Effective in high dimensional spaces.
 Still effective in cases where number of dimensions is greater than the number of samples.
 Uses a subset of training points in the decision function (called support vectors), so it is also memory efficient.
 Versatile: different Kernel functions can be specified for the decision function. It performs well under a high number of features, compared to the number of samples.*
 
 
- What are the weaknesses of the model; when does it perform poorly?
  - *If the number of features is much greater than the number of samples, avoid over-fitting in choosing Kernel functions and regularization term is crucial (this is not our case).
  SVMs do not directly provide probability estimates, these are calculated using an expensive five-fold cross-validation (since out training dataset is small, we do not worry too much about performance). It has high computation cost in training (exponentially growth in time). It can performs poorly when the number of features is much greater than the number of samples. *
  
  
- What makes this model a good candidate for the problem, given what you know about the data?
  - *It can avoid overfitting, since its very effective in high dimensional spaces*


**Ensemble method (Adaboost)**
- Describe one real-world application in industry where the model can be applied.
  - *Predicting customer churn (the propensity of customers to cease doing business with a company in a given time period) https://www.cs.rit.edu/~rlaz/PatternRecognition/slides/churn_adaboost.pdf*
  
  
- What are the strengths of the model; when does it perform well?
  - *Powerfull and successful in  a wide range of different fields (versatile); simple (require little tweaking of parameters or settings),can be less susceptible to the overfitting problem than most learning algorithms. Given that, is  seems to perform well under small datasets.*
  
  
- What are the weaknesses of the model; when does it perform poorly?
  - *AdaBoost can be sensitive to noisy data and outliers. So specially in the case when there is a small dataset mixed with noise and outliers in the samples, it can perform poorly.*
  
  
- What makes this model a good candidate for the problem, given what you know about the data?
  - *Less susceptible to the overfitting problem due to high number of features.*

**Logistic regression**
- Describe one real-world application in industry where the model can be applied.
  - *Predicting bankruptcy in the hospitality industry: https://scholarworks.umass.edu/cgi/viewcontent.cgi?&article=1152&context=jhfm*
  
  
- What are the strengths of the model; when does it perform well? 
  - *Its outputs have a nice probabilistic interpretation, and the algorithm can be regularized to avoid overfitting. Logistic models can be updated easily with new data using stochastic gradient descent. It performs well when the relationship between the features are not too complex, preferentially linear.*
  
  
- What are the weaknesses of the model; when does it perform poorly?
  - *Logistic regression tends to underperform when there are multiple or non-linear decision boundaries. They are not flexible enough to naturally capture more complex relationships. Given that, it can perform poorly when the features relate into each other in a complex and non linear form.*
  
  
- What makes this model a good candidate for the problem, given what you know about the data?
  - *Our data seems not to have to complex relationships between its features (seems linear), and with regularization we can avoid possible overfitting.*





### Setup
Run the code cell below to initialize three helper functions which you can use for training and testing the three supervised learning models you've chosen above. The functions are as follows:
- `train_classifier` - takes as input a classifier and training data and fits the classifier to the data.
- `predict_labels` - takes as input a fit classifier, features, and a target labeling and makes predictions using the F<sub>1</sub> score.
- `train_predict` - takes as input a classifier, and the training and testing data, and performs `train_clasifier` and `predict_labels`.
 - This function will report the F<sub>1</sub> score for both the training and testing data separately.

In [41]:
def train_classifier(clf, X_train, y_train):
    ''' Fits a classifier to the training data. '''
    
    # Start the clock, train the classifier, then stop the clock
    start = time()
    clf.fit(X_train, y_train)
    end = time()
    
    # Print the results
    print "Trained model in {:.4f} seconds".format(end - start)

    
def predict_labels(clf, features, target):
    ''' Makes predictions using a fit classifier based on F1 score. '''
    
    # Start the clock, make predictions, then stop the clock
    start = time()
    y_pred = clf.predict(features)
    end = time()
    
    # Print and return results
    print "Made predictions in {:.4f} seconds.".format(end - start)
    return f1_score(target.values, y_pred, pos_label='yes')


def train_predict(clf, X_train, y_train, X_test, y_test):
    ''' Train and predict using a classifer based on F1 score. '''
    
    # Indicate the classifier and the training set size
    print "Training a {} using a training set size of {}. . .".format(clf.__class__.__name__, len(X_train))
    
    # Train the classifier
    train_classifier(clf, X_train, y_train)
    
    # Print the results of prediction for both training and testing
    print "F1 score for training set: {:.4f}.".format(predict_labels(clf, X_train, y_train))
    print "F1 score for test set: {:.4f}.".format(predict_labels(clf, X_test, y_test))

### Implementation: Model Performance Metrics
With the predefined functions above, you will now import the three supervised learning models of your choice and run the `train_predict` function for each one. Remember that you will need to train and predict on each classifier for three different training set sizes: 100, 200, and 300. Hence, you should expect to have 9 different outputs below — 3 for each model using the varying training set sizes. In the following code cell, you will need to implement the following:
- Import the three supervised learning models you've discussed in the previous section.
- Initialize the three models and store them in `clf_A`, `clf_B`, and `clf_C`.
 - Use a `random_state` for each model you use, if provided.
 - **Note:** Use the default settings for each model — you will tune one specific model in a later section.
- Create the different training set sizes to be used to train each model.
 - *Do not reshuffle and resplit the data! The new training points should be drawn from `X_train` and `y_train`.*
- Fit each model with each training set size and make predictions on the test set (9 in total).  
**Note:** Three tables are provided after the following code cell which can be used to store your results.

In [50]:
# TODO: Import the three supervised learning models from sklearn
from sklearn.ensemble import AdaBoostClassifier
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression

# TODO: Initialize the three models
clf_A = AdaBoostClassifier(random_state=seed)
clf_B = LinearSVC(random_state=seed)
clf_C = LogisticRegression(random_state=seed)

# TODO: Set up the training set sizes
X_train_100 = X_train[0:100]
y_train_100 = y_train[0:100]

X_train_200 = X_train[100:300]
y_train_200 = y_train[100:300]

X_train_300 = X_train[:300]
y_train_300 = y_train[:300]

# TODO: Execute the 'train_predict' function for each classifier and each training set size
# train_predict(clf, X_train, y_train, X_test, y_test)

# AdaBoostClassifier
train_predict(clf_A, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_A, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_A, X_train_300, y_train_300, X_test, y_test)
# Support Vector Machine
train_predict(clf_B, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_B, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_B, X_train_300, y_train_300, X_test, y_test)
# LogisticRegression
train_predict(clf_C, X_train_100, y_train_100, X_test, y_test)
train_predict(clf_C, X_train_200, y_train_200, X_test, y_test)
train_predict(clf_C, X_train_300, y_train_300, X_test, y_test)

Training a AdaBoostClassifier using a training set size of 100. . .
Trained model in 0.0839 seconds
Made predictions in 0.0037 seconds.
F1 score for training set: 0.9718.
Made predictions in 0.0037 seconds.
F1 score for test set: 0.8219.
Training a AdaBoostClassifier using a training set size of 200. . .
Trained model in 0.0855 seconds
Made predictions in 0.0044 seconds.
F1 score for training set: 0.8712.
Made predictions in 0.0035 seconds.
F1 score for test set: 0.7971.
Training a AdaBoostClassifier using a training set size of 300. . .
Trained model in 0.0907 seconds
Made predictions in 0.0048 seconds.
F1 score for training set: 0.8441.
Made predictions in 0.0036 seconds.
F1 score for test set: 0.8000.
Training a LinearSVC using a training set size of 100. . .
Trained model in 0.0069 seconds
Made predictions in 0.0002 seconds.
F1 score for training set: 0.9167.
Made predictions in 0.0002 seconds.
F1 score for test set: 0.6866.
Training a LinearSVC using a training set size of 200. . 

### Tabular Results
Edit the cell below to see how a table can be designed in [Markdown](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#tables). You can record your results from above in the tables provided.

** Classifer 1 - AdaBoostClassifier**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |        0.0855           |      0.0039            |   0.9718         |    0.8219       |
| 200               |        0.0884           |      0.0039            |   0.8712         |    0.7971       |
| 300               |        0.0916           |      0.0036            |   0.8441         |    0.8000       |

** Classifer 2 - LinearSVC**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |     0.0069              |       0.0002           |    0.9167        |     0.6866      |
| 200               |     0.0140              |       0.0002           |    0.8379        |     0.7634      |
| 300               |     0.0234              |       0.0002           |    0.8254        |     0.8054      |

** Classifer 3 - LogisticRegression**  

| Training Set Size | Training Time | Prediction Time (test) | F1 Score (train) | F1 Score (test) |
| :---------------: | :---------------------: | :--------------------: | :--------------: | :-------------: |
| 100               |   0.0011                |    0.0002              |   0.9128         |    0.7692       |
| 200               |   0.0014                |    0.0002              |   0.8444         |   0.7914        |
| 300               |   0.0022                |    0.0002              |   0.8369         |   0.7832        | 

## Choosing the Best Model
In this final section, you will choose from the three supervised learning models the *best* model to use on the student data. You will then perform a grid search optimization for the model over the entire training set (`X_train` and `y_train`) by tuning at least one parameter to improve upon the untuned model's F<sub>1</sub> score. 

### Question 3 - Choosing the Best Model
*Based on the experiments you performed earlier, in one to two paragraphs, explain to the board of supervisors what single model you chose as the best model. Which model is generally the most appropriate based on the available data, limited resources, cost, and performance?*

**Answer: ** 
Based on the avaliable format of data, the absolute computacional cost and performance, the best option seems to be the Support Vector Machine with a linear kernel. Its training time seems big O(n²) to the number of samples, which is not costly, considering that the absolute time of training (only 0.0234 seconds in an average personal computer) is small (even considering scalling to 100, 1000 times the original dataset size). It generalizes well, becoming better with more training samples and preventing overfitting (even in the presence of a high number of features). And last but not least, it has the best F1 Score in the testing set of all the three models. 

### Question 4 - Model in Layman's Terms
*In one to two paragraphs, explain to the board of directors in layman's terms how the final model chosen is supposed to work. Be sure that you are describing the major qualities of the model, such as how the model is trained and how the model makes a prediction. Avoid using advanced mathematical or technical jargon, such as describing equations or discussing the algorithm implementation.*

**Answer: **

Given the problem exposed, i.e, identify students who might need early intervention before they fail to graduate, we have chosen a final model to solve this issue. We made it work by using past information from a students database (age, sex, activities,passed) as input to train our model. After trained, we can use the model, by inputting data from a new student in it, and receiving the answer if we might or not intervent in this new case.

Our model is robust, fairly easy and fast to train and retrain (when adding more data to our students dataset), and very fast to return a result after trained. It can give a result with around 80% of confidence, so we can improve the chances of helping problematic students by reaching them before a highly probable failing.

### Implementation: Model Tuning
Fine tune the chosen model. Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. You will need to use the entire training set for this. In the code cell below, you will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
- Initialize the classifier you've chosen and store it in `clf`.
- Create the F<sub>1</sub> scoring function using `make_scorer` and store it in `f1_scorer`.
 - Set the `pos_label` parameter to the correct value!
- Perform grid search on the classifier `clf` using `f1_scorer` as the scoring method, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_obj`.

In [65]:
# TODO: Import 'GridSearchCV' and 'make_scorer'
from sklearn.grid_search import GridSearchCV
from sklearn.metrics import make_scorer
# TODO: Create the parameters list you wish to tune
parameters = {'C':[0.1, 1, 10, 100, 1000, 10000],
              'max_iter':[10, 100, 1000, 10000] ,
              'loss': ['hinge', 'squared_hinge']}

# TODO: Initialize the classifier
clf = LinearSVC(random_state=seed)

# TODO: Make an f1 scoring function using 'make_scorer' 
f1_scorer = make_scorer(f1_score, pos_label='yes')

# TODO: Perform grid search on the classifier using the f1_scorer as the scoring method
grid_obj = GridSearchCV(clf, parameters, scoring=f1_scorer)

# TODO: Fit the grid search object to the training data and find the optimal parameters
grid_obj = grid_obj.fit(X_train, y_train)

# Get the estimator
clf = grid_obj.best_estimator_

# Report the final F1 score for training and testing after parameter tuning
print "Tuned model has a training F1 score of {:.4f}.".format(predict_labels(clf, X_train, y_train))
print "Tuned model has a testing F1 score of {:.4f}.".format(predict_labels(clf, X_test, y_test))

Made predictions in 0.0047 seconds.
Tuned model has a training F1 score of 0.8090.
Made predictions in 0.0003 seconds.
Tuned model has a testing F1 score of 0.8228.


### Question 5 - Final F<sub>1</sub> Score
*What is the final model's F<sub>1</sub> score for training and testing? How does that score compare to the untuned model?*

**Answer: **
The training F1 score went from 0.8254 to 0.8090(~2% worse), but the testing F1 score went from 0.8054 to 0.8228 (~2% better). Since our main concern is increasing the F1 score on the testing set (generalizing better), our tuned model became better at generalizing than the previous untuned model (tuning worked).

> **Note**: Once you have completed all of the code implementations and successfully answered each question above, you may finalize your work by exporting the iPython Notebook as an HTML document. You can do this by using the menu above and navigating to  
**File -> Download as -> HTML (.html)**. Include the finished document along with this notebook as your submission.